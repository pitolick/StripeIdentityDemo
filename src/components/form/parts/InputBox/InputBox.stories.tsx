import type { Meta, StoryObj } from '@storybook/react'
import { InputBox } from './InputBox'

const meta: Meta<typeof InputBox> = {
  title: 'Form/Parts/InputBox',
  component: InputBox,
  tags: ['autodocs'],
}

export default meta
type Story = StoryObj<typeof InputBox>

/**
 * 基本形
 */
export const Base: Story = {
  args: {
    nameId: 'base',
    children: <input type="text" className="w-full" />,
  },
}

/**
 * ラベル
 */
export const Label: Story = {
  args: {
    nameId: 'label',
    label: 'ラベル',
    children: <input type="text" className="w-full" />,
  },
}

/**
 * 必須
 */
export const Required: Story = {
  args: {
    nameId: 'required',
    label: '必須',
    required: true,
    children: <input type="text" className="w-full" />,
  },
}

/**
 * 任意
 */
export const Optional: Story = {
  args: {
    nameId: 'optional',
    label: '任意',
    required: false,
    children: <input type="text" className="w-full" />,
  },
}

/**
 * 注釈
 */
export const Note: Story = {
  args: {
    nameId: 'note',
    label: '注釈',
    note: { list: ['注釈です'] },
    children: <input type="text" className="w-full" />,
  },
}

/**
 * エラー
 */
export const Error: Story = {
  args: {
    nameId: 'error',
    label: 'エラー',
    required: true,
    note: { list: ['注釈です'] },
    error: { type: 'required', message: '必須項目です' },
    children: <input type="text" className="w-full" />,
  },
}

/**
 * 子要素の後ろに追加する要素
 */
export const AfterChildren: Story = {
  args: {
    nameId: 'after-children',
    label: '子要素の後ろに追加する要素',
    children: <input type="text" className="w-full" />,
    afterChildren: <span className="text-sm text-gray-500">afterChildren</span>,
  },
}

/**
 * 確認画面用
 */
export const Confirm: Story = {
  args: {
    nameId: 'confirm',
    label: '確認画面用',
    children: '確認画面用',
    styleType: 'confirm',
  },
}
