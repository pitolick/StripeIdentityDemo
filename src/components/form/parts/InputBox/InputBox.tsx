/**
 * 入力コンポーネントのレイアウトを定義する
 * @module InputBox
 */
import React from 'react'
import { Required } from '@/components/form/parts'
import { Note, NoteProps } from '@/components/elements'
import { FieldError, Merge, FieldErrorsImpl } from 'react-hook-form'

/**
 * 入力コンポーネントのプロパティ
 * @memo バリデーションスキーマ（zod）は各入力子コンポーネント内でexportしformコンポーネント内でresolverに指定する
 * @typedef {Object} InputBoxProps
 */
type NestedError = {
  [key: string]: FieldError | Merge<FieldError, FieldErrorsImpl<any>> | NestedError
}
export interface InputBoxProps {
  /**
   * フォームの名前（ID）
   */
  nameId: string
  /**
   * フォームのラベル
   */
  label?: string
  /**
   * 注釈
   */
  note?: NoteProps
  /**
   * 必須任意の表示
   */
  required?: boolean
  /**
   * useFormから取得できるエラーオブジェクト
   */
  error?: FieldError | Merge<FieldError, FieldErrorsImpl<any>> | NestedError
  /**
   * 子要素
   */
  children: React.ReactNode
  /**
   * 子要素の後ろに表示する要素
   */
  afterChildren?: React.ReactNode
  /**
   * 表示スタイルタイプ
   */
  styleType?: 'input' | 'confirm'
  /**
   * コンテンツ幅
   */
  contentWidth?: 'full' | 'auto'
}

/**
 * 入力コンポーネントのレイアウトを定義する
 * @param {InputBoxProps} props - 入力コンポーネントのプロパティ
 * @returns {React.FC} - 入力コンポーネント
 */
export const InputBox: React.FC<InputBoxProps> = (props) => {
  const { nameId, label, note, required, error, children, afterChildren, styleType, contentWidth = 'auto' } = props
  return (
    <div className={`${contentWidth === 'full' && 'w-full flex-auto'} space-y-2 ${styleType === 'confirm' && 'border-b border-b-gray-700'}`}>
      {(required !== undefined || label) && (
        <div className="flex items-center justify-start gap-x-2">
          {/* requiredがundefinedの時は非表示 */}
          {required !== undefined && <Required required={required} />}
          {/* labelが指定されていないときは非表示 */}
          {label && (
            <label htmlFor={nameId} className="block text-sm font-bold">
              {label}
            </label>
          )}
        </div>
      )}
      <div className={`space-y-2 ${styleType === 'confirm' && 'px-2 pb-3 pt-2'}`}>
        <div className={`flex w-full items-center justify-start gap-x-3`}>
          {children}
          {afterChildren}
        </div>
        {/* 注釈 */}
        {note && <Note {...note} />}
        {/* エラーメッセージ */}
        {error && (
          <p className="text-sm text-red">
            <>{error?.message}</>
          </p>
        )}
      </div>
    </div>
  )
}
