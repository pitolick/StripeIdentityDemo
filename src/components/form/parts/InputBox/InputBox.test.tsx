import React from 'react'
import { render } from '@testing-library/react'
import { InputBox, InputBoxProps } from './InputBox'

describe('InputBox', () => {
  const defaultProps: InputBoxProps = {
    nameId: 'test',
    label: 'テスト',
    note: { list: ['注釈'] },
    required: true,
    error: { type: 'test', message: 'エラーメッセージ' },
    children: <input type="text" />,
    afterChildren: <button>ボタン</button>,
    styleType: 'input',
    contentWidth: 'full',
  }

  it('正しくレンダリングされること', () => {
    const { container } = render(<InputBox {...defaultProps} />)
    expect(container.firstChild).toMatchSnapshot()
  })

  it('styleTypeがinputの場合、正しくレンダリングされること', () => {
    const { container } = render(<InputBox {...defaultProps} styleType="input" />)
    expect(container.firstChild).toMatchSnapshot()
  })

  it('styleTypeがconfirmの場合、正しくレンダリングされること', () => {
    const { container } = render(<InputBox {...defaultProps} styleType="confirm" />)
    expect(container.firstChild).toMatchSnapshot()
  })

  it('requiredがfalseの場合、必須任意の表示が非表示になること', () => {
    const { container } = render(<InputBox {...defaultProps} required={false} />)
    expect(container.firstChild).toMatchSnapshot()
  })

  it('contentWidthがautoの場合、コンテンツ幅がautoになること', () => {
    const { container } = render(<InputBox {...defaultProps} contentWidth="auto" />)
    expect(container.firstChild).toMatchSnapshot()
  })

  it('contentWidthが未入力の場合、コンテンツ幅がデフォルト値になること', () => {
    const { container } = render(<InputBox {...defaultProps} contentWidth={undefined} />)
    expect(container.firstChild).toMatchSnapshot()
  })

  it('labelが未入力の場合、ラベルが非表示になること', () => {
    const { container } = render(<InputBox {...defaultProps} label={undefined} />)
    expect(container.firstChild).toMatchSnapshot()
  })

  it('noteが未入力の場合、注釈が非表示になること', () => {
    const { container } = render(<InputBox {...defaultProps} note={undefined} />)
    expect(container.firstChild).toMatchSnapshot()
  })

  it('errorが未入力の場合、エラーが非表示になること', () => {
    const { container } = render(<InputBox {...defaultProps} error={undefined} />)
    expect(container.firstChild).toMatchSnapshot()
  })
})
