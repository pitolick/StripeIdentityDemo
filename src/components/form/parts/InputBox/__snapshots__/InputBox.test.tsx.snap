// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputBox contentWidthがautoの場合、コンテンツ幅がautoになること 1`] = `
<div
  class="false space-y-2 false"
>
  <div
    class="flex items-center justify-start gap-x-2"
  >
    <span
      class="text-xs text-center px-1.5 md:px-3 py-1 rounded-md whitespace-nowrap bg-red text-white"
    >
      必須
    </span>
    <label
      class="block text-sm font-bold"
      for="test"
    >
      テスト
    </label>
  </div>
  <div
    class="space-y-2 false"
  >
    <div
      class="flex w-full items-center justify-start gap-x-3"
    >
      <input
        type="text"
      />
      <button>
        ボタン
      </button>
    </div>
    <ul>
      <li
        class="text-xs text-gray-700 before:me-1 ps-4 -indent-4 before:content-['※']"
      >
        注釈
      </li>
    </ul>
    <p
      class="text-sm text-red"
    >
      エラーメッセージ
    </p>
  </div>
</div>
`;

exports[`InputBox contentWidthが未入力の場合、コンテンツ幅がデフォルト値になること 1`] = `
<div
  class="false space-y-2 false"
>
  <div
    class="flex items-center justify-start gap-x-2"
  >
    <span
      class="text-xs text-center px-1.5 md:px-3 py-1 rounded-md whitespace-nowrap bg-red text-white"
    >
      必須
    </span>
    <label
      class="block text-sm font-bold"
      for="test"
    >
      テスト
    </label>
  </div>
  <div
    class="space-y-2 false"
  >
    <div
      class="flex w-full items-center justify-start gap-x-3"
    >
      <input
        type="text"
      />
      <button>
        ボタン
      </button>
    </div>
    <ul>
      <li
        class="text-xs text-gray-700 before:me-1 ps-4 -indent-4 before:content-['※']"
      >
        注釈
      </li>
    </ul>
    <p
      class="text-sm text-red"
    >
      エラーメッセージ
    </p>
  </div>
</div>
`;

exports[`InputBox errorが未入力の場合、エラーが非表示になること 1`] = `
<div
  class="w-full flex-auto space-y-2 false"
>
  <div
    class="flex items-center justify-start gap-x-2"
  >
    <span
      class="text-xs text-center px-1.5 md:px-3 py-1 rounded-md whitespace-nowrap bg-red text-white"
    >
      必須
    </span>
    <label
      class="block text-sm font-bold"
      for="test"
    >
      テスト
    </label>
  </div>
  <div
    class="space-y-2 false"
  >
    <div
      class="flex w-full items-center justify-start gap-x-3"
    >
      <input
        type="text"
      />
      <button>
        ボタン
      </button>
    </div>
    <ul>
      <li
        class="text-xs text-gray-700 before:me-1 ps-4 -indent-4 before:content-['※']"
      >
        注釈
      </li>
    </ul>
  </div>
</div>
`;

exports[`InputBox labelが未入力の場合、ラベルが非表示になること 1`] = `
<div
  class="w-full flex-auto space-y-2 false"
>
  <div
    class="flex items-center justify-start gap-x-2"
  >
    <span
      class="text-xs text-center px-1.5 md:px-3 py-1 rounded-md whitespace-nowrap bg-red text-white"
    >
      必須
    </span>
  </div>
  <div
    class="space-y-2 false"
  >
    <div
      class="flex w-full items-center justify-start gap-x-3"
    >
      <input
        type="text"
      />
      <button>
        ボタン
      </button>
    </div>
    <ul>
      <li
        class="text-xs text-gray-700 before:me-1 ps-4 -indent-4 before:content-['※']"
      >
        注釈
      </li>
    </ul>
    <p
      class="text-sm text-red"
    >
      エラーメッセージ
    </p>
  </div>
</div>
`;

exports[`InputBox noteが未入力の場合、注釈が非表示になること 1`] = `
<div
  class="w-full flex-auto space-y-2 false"
>
  <div
    class="flex items-center justify-start gap-x-2"
  >
    <span
      class="text-xs text-center px-1.5 md:px-3 py-1 rounded-md whitespace-nowrap bg-red text-white"
    >
      必須
    </span>
    <label
      class="block text-sm font-bold"
      for="test"
    >
      テスト
    </label>
  </div>
  <div
    class="space-y-2 false"
  >
    <div
      class="flex w-full items-center justify-start gap-x-3"
    >
      <input
        type="text"
      />
      <button>
        ボタン
      </button>
    </div>
    <p
      class="text-sm text-red"
    >
      エラーメッセージ
    </p>
  </div>
</div>
`;

exports[`InputBox requiredがfalseの場合、必須任意の表示が非表示になること 1`] = `
<div
  class="w-full flex-auto space-y-2 false"
>
  <div
    class="flex items-center justify-start gap-x-2"
  >
    <span
      class="text-xs text-center px-1.5 md:px-3 py-1 rounded-md whitespace-nowrap bg-gray-500 text-white"
    >
      任意
    </span>
    <label
      class="block text-sm font-bold"
      for="test"
    >
      テスト
    </label>
  </div>
  <div
    class="space-y-2 false"
  >
    <div
      class="flex w-full items-center justify-start gap-x-3"
    >
      <input
        type="text"
      />
      <button>
        ボタン
      </button>
    </div>
    <ul>
      <li
        class="text-xs text-gray-700 before:me-1 ps-4 -indent-4 before:content-['※']"
      >
        注釈
      </li>
    </ul>
    <p
      class="text-sm text-red"
    >
      エラーメッセージ
    </p>
  </div>
</div>
`;

exports[`InputBox styleTypeがconfirmの場合、正しくレンダリングされること 1`] = `
<div
  class="w-full flex-auto space-y-2 border-b border-b-gray-700"
>
  <div
    class="flex items-center justify-start gap-x-2"
  >
    <span
      class="text-xs text-center px-1.5 md:px-3 py-1 rounded-md whitespace-nowrap bg-red text-white"
    >
      必須
    </span>
    <label
      class="block text-sm font-bold"
      for="test"
    >
      テスト
    </label>
  </div>
  <div
    class="space-y-2 px-2 pb-3 pt-2"
  >
    <div
      class="flex w-full items-center justify-start gap-x-3"
    >
      <input
        type="text"
      />
      <button>
        ボタン
      </button>
    </div>
    <ul>
      <li
        class="text-xs text-gray-700 before:me-1 ps-4 -indent-4 before:content-['※']"
      >
        注釈
      </li>
    </ul>
    <p
      class="text-sm text-red"
    >
      エラーメッセージ
    </p>
  </div>
</div>
`;

exports[`InputBox styleTypeがinputの場合、正しくレンダリングされること 1`] = `
<div
  class="w-full flex-auto space-y-2 false"
>
  <div
    class="flex items-center justify-start gap-x-2"
  >
    <span
      class="text-xs text-center px-1.5 md:px-3 py-1 rounded-md whitespace-nowrap bg-red text-white"
    >
      必須
    </span>
    <label
      class="block text-sm font-bold"
      for="test"
    >
      テスト
    </label>
  </div>
  <div
    class="space-y-2 false"
  >
    <div
      class="flex w-full items-center justify-start gap-x-3"
    >
      <input
        type="text"
      />
      <button>
        ボタン
      </button>
    </div>
    <ul>
      <li
        class="text-xs text-gray-700 before:me-1 ps-4 -indent-4 before:content-['※']"
      >
        注釈
      </li>
    </ul>
    <p
      class="text-sm text-red"
    >
      エラーメッセージ
    </p>
  </div>
</div>
`;

exports[`InputBox 正しくレンダリングされること 1`] = `
<div
  class="w-full flex-auto space-y-2 false"
>
  <div
    class="flex items-center justify-start gap-x-2"
  >
    <span
      class="text-xs text-center px-1.5 md:px-3 py-1 rounded-md whitespace-nowrap bg-red text-white"
    >
      必須
    </span>
    <label
      class="block text-sm font-bold"
      for="test"
    >
      テスト
    </label>
  </div>
  <div
    class="space-y-2 false"
  >
    <div
      class="flex w-full items-center justify-start gap-x-3"
    >
      <input
        type="text"
      />
      <button>
        ボタン
      </button>
    </div>
    <ul>
      <li
        class="text-xs text-gray-700 before:me-1 ps-4 -indent-4 before:content-['※']"
      >
        注釈
      </li>
    </ul>
    <p
      class="text-sm text-red"
    >
      エラーメッセージ
    </p>
  </div>
</div>
`;
